Adding Restore URL support to the new OM restore dialog for MI Server

ssreerama opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

Adds the ability to support the restore database from URL, for a new restore dialog and for MI databases.

Note: TODO: Tasks tab shows multiple messages (unwanted one even in fail cases) will send a new PR if it takes time.

![NewRestoreURL_MI](https://github.com/microsoft/azuredatastudio/assets/74571829/fbcd5f86-075d-4678-ac8c-4420eba54900)

corivera commented on extensions/mssql/src/objectManagement/utils.ts:
```diff
@@ -54,3 +54,29 @@ export function convertNumToTwoDecimalStringInMB(value: number): string {
+/**
+ * Backup media device type: https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.devicetype
+ */
+export enum MediaDeviceType {
```
The constants.ts file in the same folder would probably be a better place for these.

corivera commented on extensions/mssql/src/objectManagement/ui/restoreDatabaseDialog.ts:
```diff
@@ -179,16 +201,18 @@ export class RestoreDatabaseDialog extends ObjectManagementDialogBase<Database,
 		const isRestoreFromBackupFile = this.restoreFrom.value === localizedConstants.RestoreFromBackupFileOptionText;
 		let options: { [key: string]: any } = {};
-		Object.entries(this.objectInfo.restorePlanResponse.planDetails).forEach(([key, value]) => {
+		Object.entries(this.objectInfo.restorePlanResponse?.planDetails).forEach(([key, value]) => {
```
Looks like Object.entries throws an error if it's passed an undefined value. Could you add a check for that here?

ssreerama replied:
Issue was caused by intercahnging the restroeFrom dd option and it is resolved now.

corivera commented on extensions/mssql/src/objectManagement/ui/restoreDatabaseDialog.ts:
```diff
@@ -201,17 +225,21 @@ export class RestoreDatabaseDialog extends ObjectManagementDialogBase<Database,
+		this.isManaged = this.viewInfo.isManagedInstance;
+
 		// Restore from
+		const restoreFromDdOptions = this.isManaged ? [localizedConstants.RestoreFromUrlText] : [localizedConstants.RestoreFromDatabaseOptionText, localizedConstants.RestoreFromBackupFileOptionText];
```
```suggestion
		const restoreFromDbOptions = this.isManaged ? [localizedConstants.RestoreFromUrlText] : [localizedConstants.RestoreFromDatabaseOptionText, localizedConstants.RestoreFromBackupFileOptionText];
```

Charles-Gagnon commented on extensions/mssql/src/objectManagement/ui/restoreDatabaseDialog.ts:
```diff
@@ -53,6 +55,8 @@ export class RestoreDatabaseDialog extends ObjectManagementDialogBase<Database,
 	private logFileFolderButton: azdata.ButtonComponent;
 	private logFileFolderContainer: azdata.FlexContainer;
+	private targetDatabase: azdata.DropDownComponent;
+	private isManaged: boolean;
```
I'd suggest the full `isManagedInstance` to make it clear when this would be true

corivera commented:
Tried the changes out locally since they were a bit complicated to parse through just in text. I'm hitting an issue where the dialog is stuck in a spinning state after selecting a backup file from blob storage to use for the Restore from URL scenario. I'm using both of your ADS and STS branches for this change.

Looks like some `planDetails `fields are undefined when I'm hitting that problem. So like when trying to get the `currentValue `for the `sourceDatabaseName `or `targetDatabaseName  `in the `planDetails`.

For my test scenario, I made a backup from a 2022 Docker instance, and then tried to select that backup file from a storage account when restoring to a newly created Managed Instance database. The backup file might end up being incompatible due to version differences, but I'm not even able to get to the step where I can Script out the operation or hit Restore. The buttons remain grayed out.

ssreerama commented:
Hey @corivera , I tried to repro the issue and couldn't, it is working fine. Below are the steps that I followed.
1. Created a database and added a table with data in 2022 Docker instance.
2. Using SSMS, I took a backup to URL. 
3. In Managed Instance Server, opened the restore database and selected the above backed up file from the blob storage.
4. It took few seconds(hardly 4-6) to load the restore plan with the selected one.
5. Selected a new database name (as we cannot restore on existing DB in MI). 
6. both script and restore button worked and I can see the new database.

Please correct me if I missed any step. Thanks

ssreerama commented:
Able to found couple of issues with Docker and fixed them also, added URL support to on-prem servers. (MI supports only URL, all other supports database/back-up files/ URL)

corivera commented on extensions/mssql/src/objectManagement/ui/restoreDatabaseDialog.ts:
```diff
@@ -108,26 +112,30 @@ export class RestoreDatabaseDialog extends ObjectManagementDialogBase<Database,
-			], false)
-		};
-		tabs.push(this.optionsTab);
+		if (!this.isManagedInstance) {
```
You can combine this with the if-case above it, since they have the same if-condition.

ssreerama commented on extensions/mssql/src/objectManagement/constants.ts:
```diff
@@ -69,3 +69,29 @@ export const enum TelemetryActions {
+/**
+ * Backup media device type: https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.devicetype
+ */
+export enum MediaDeviceType {
```
Yes, was copied these to mssql backupDatabaseDialog.ts and moved it to constant file to utilize for restore as well. @corivera was there a reason to use copy from here: https://github.com/Microsoft/azuredatastudio/blob/main/src/sql/workbench/common/backup/constants.ts ??

corivera replied:
@Charles-Gagnon  Are we able to import core files in extensions?

Charles-Gagnon replied:
Oh whoops, got my wires crossed. Was looking at another file at the same time and thought this was in the MSSQL extension. So nevermind then, what you have is good!

(no you can't)
